### PR TITLE
[main][feat]: create alpacafeeder02 contract for supporting ftm.

### PR DESCRIPTION
**Pull request by blackalpacaa**
## Description

- Created contract AlpacaFeeder02 including unit test and integration test for supporting FTM chain.

### CODE REVIEW

File: contracts/8.10/mocks/MockMiniFL.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    if (proxyToken.myBalance() > 0) {
+      SafeToken.safeApprove(proxyToken, _for, proxyToken.myBalance());
+      proxyToken.safeTransfer(_for, proxyToken.myBalance());
+      SafeToken.safeApprove(proxyToken, _for, 0);
+    }
+  }
+
+  function addPool(address _stakeToken) external {
+    // do nothing
```
Comment (alpanaca) [inline contracts/8.10/mocks/MockMiniFL.sol:64]: Remove this

File: tests/helpers/timelock.ts [hunk 2]
```diff
@@ -1,7 +1,9 @@
-import { FairLaunch, Timelock, Timelock__factory } from "@alpaca-finance/alpaca-contract/typechain";
+import { FairLaunch, Timelock, Timelock__factory, MiniFL } from "@alpaca-finance/alpaca-contract/typechain";
 import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { zeroAddress } from "ethereumjs-util";
 import { ethers, BigNumber } from "ethers";
-import * as addresses from "../constants/addresses";
+import * as addresses from "../constants/bsc/addresses";
+import * as ftmAddresses from "../constants/bsc/addresses";
```
Comment (alpanaca) [inline tests/helpers/timelock.ts:6]:
```suggestion
import * as ftmAddresses from "../constants/ftm/addresses";
```

File: contracts/8.10/AlpacaFeeder02.sol [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    token = _token;
+    proxyToken = _proxyToken;
+    miniFLPoolId = _miniFLPoolId;
+    miniFL = IMiniFL(_miniFLAddress);
+    grassHouse = IGrassHouse(_grasshouseAddress);
+
+    address _stakeToken = address(miniFL.stakingToken(_miniFLPoolId));
+
+    require(_stakeToken == _proxyToken, "!same stakeToken");
```
Comment (jr-alpaca) [inline contracts/8.10/AlpacaFeeder02.sol:67]: Please use revert Error
```suggestion
    if(_stakeToken != _proxyToken) revert AlpacaFeeder02_InvalidToken();

```

File: contracts/8.10/mocks/MockMiniFL.sol [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  ξ ξ ξ~～~ξ ξ ξ 
+　 ξ_ξξ_ξ　ξ_ξξ_ξ
+Alpaca Fin Corporation
+*/
+
+pragma solidity 0.8.10;
+
+import "../interfaces/IMiniFL.sol";
+import "../SafeToken.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
```
Comment (jr-alpaca) [inline contracts/8.10/mocks/MockMiniFL.sol:18]:
```suggestion
import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";

import "../interfaces/IMiniFL.sol";

import "../SafeToken.sol";

```

File: contracts/8.10/AlpacaFeeder02.sol [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    grassHouse.feed(_feedAmount);
+    emit LogFeedGrassHouse(_feedAmount);
+  }
+
+  /// @notice Set a new GrassHouse
+  /// @param _newGrassHouse - new GrassHouse address
+  function setGrassHouse(IGrassHouse _newGrassHouse) external onlyOwner {
+    address _prevGrassHouse = address(grassHouse);
+    grassHouse = _newGrassHouse;
+    emit LogSetNewGrassHouse(msg.sender, _prevGrassHouse, address(_newGrassHouse));
```
Comment (sora-alpaca) [inline contracts/8.10/AlpacaFeeder02.sol:114]:
```suggestion
    emit LogSetGrassHouse(msg.sender, _prevGrassHouse, address(_newGrassHouse));
```

File: contracts/8.10/AlpacaFeeder02.sol [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+contract AlpacaFeeder02 is IVault, Initializable, OwnableUpgradeable {
+  /// @notice Libraries
+  using SafeToken for address;
+
+  /// @notice Events
+  event LogFeedGrassHouse(uint256 _feedAmount);
+  event LogMiniFLDeposit();
+  event LogMiniFLWithdraw();
+  event LogMiniFLHarvest(address _caller, uint256 _harvestAmount);
+  event LogSetNewGrassHouse(address indexed _caller, address _prevGrassHouse, address _newGrassHouse);
```
Comment (sora-alpaca) [inline contracts/8.10/AlpacaFeeder02.sol:37]:
```suggestion
  event LogSetGrassHouse(address indexed _caller, address _prevGrassHouse, address _newGrassHouse);
```

File: tests/helpers/deploy.ts [hunk 7]
```diff
@@ -10,8 +10,11 @@ import {
   GrassHouse__factory,
   AlpacaFeeder,
   AlpacaFeeder__factory,
+  AlpacaFeeder02__factory,
+  AlpacaFeeder02,
 } from "../../typechain";
-import * as addresses from "../constants/addresses";
+import * as addresses from "../constants/bsc/addresses";
```
Comment (sora-alpaca) [inline tests/helpers/deploy.ts:16]:
```suggestion
import * as bscAddresses from "../constants/bsc/addresses";
```

File: tests/helpers/timelock.ts [hunk 8]
```diff
@@ -1,7 +1,9 @@
-import { FairLaunch, Timelock, Timelock__factory } from "@alpaca-finance/alpaca-contract/typechain";
+import { FairLaunch, Timelock, Timelock__factory, MiniFL } from "@alpaca-finance/alpaca-contract/typechain";
 import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { zeroAddress } from "ethereumjs-util";
 import { ethers, BigNumber } from "ethers";
-import * as addresses from "../constants/addresses";
+import * as addresses from "../constants/bsc/addresses";
```
Comment (sora-alpaca) [inline tests/helpers/timelock.ts:5]:
```suggestion
import * as bscAddresses from "../constants/bsc/addresses";
```

File: tests/unit/AlpacaFeeder02.test.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    // Assign contract signer
+    alpacaAsAlice = BEP20__factory.connect(alpaca.address, alice);
+    feederAsAlice = AlpacaFeeder02__factory.connect(feeder.address, alice);
+  }
+
+  beforeEach(async () => {
+    await waffle.loadFixture(fixture);
+  });
+
+  context("#initialize", () => {
```
Comment (sora-alpaca) [inline tests/unit/AlpacaFeeder02.test.ts:95]: add testcase for `#setGrasshouse` function

**Review by sora-alpaca — CHANGES_REQUESTED** [review]

**Review by jr-alpaca — APPROVED** [review]

File: tests/unit/AlpacaFeeder02.test.ts [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+      it("should continue feed", async () => {
+        await alpacaAsAlice.transfer(feeder.address, ethers.utils.parseEther("20"));
+        expect(await alpaca.balanceOf(feeder.address)).to.be.eq(ethers.utils.parseEther("20"));
+        await expect(feeder.feedGrassHouse()).to.be.emit(feeder, "LogFeedGrassHouse");
+        expect(await alpaca.balanceOf(feeder.address)).to.be.eq(ethers.utils.parseEther("0"));
+        expect(await alpaca.balanceOf(grassHouse.address)).to.be.eq(ethers.utils.parseEther("20"));
+      });
+    });
+  });
+  describe("#setGrassHouse", async () => {
```
Comment (sora-alpaca) [inline tests/unit/AlpacaFeeder02.test.ts:217]: 1 enter

**Review by sora-alpaca — CHANGES_REQUESTED** [review]

File: tests/helpers/timelock.ts [hunk 11]
```diff
@@ -37,3 +39,43 @@ export class TimelockHelper {
     return poolLength.sub(BigNumber.from(1));
   }
 }
+
```
Comment (sora-alpaca) [inline tests/helpers/timelock.ts:42]: add TODO: uncomment if miniFL owner is timelock else remove this

**Review by sora-alpaca — APPROVED** [review]

**Review by alpanaca — APPROVED** [review]

**Pull request merged**